packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp

```diff
@@ -105,8 +105,46 @@ void jacobiSetup(RCP<Teuchos::ParameterList> params,
     // extract inverse of diagonal from matrix
     diag[j] = VectorFactory::Build(regionGrpMats[j]->getRowMap(), true);
     regionGrpMats[j]->getLocalDiagCopy(*diag[j]);
-    diag[j]->elementWiseMultiply(SC_ONE, *diag[j], *regionInterfaceScaling[j], SC_ZERO); // ToDo Does it work to pass in diag[j], but also return into the same variable?
+    diag[j]->elementWiseMultiply(SC_ONE, *regionInterfaceScaling[j], *diag[j], SC_ZERO); // ToDo Does it work to pass in diag[j], but also return into the same variable?
     diag[j]->reciprocal(*diag[j]);
+
+    // RCP<Teuchos::FancyOStream> out = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    // regionInterfaceScaling[j]->describe(*out, Teuchos::VERB_EXTREME);
+    // diag[j]->describe(*out, Teuchos::VERB_EXTREME);
+  }
+
+  params->set<Teuchos::Array<RCP<Vector> > >("jacobi: inverse diagonal", diag);
+}
+
+/*! \brief performs Jacobi setup
+ *
+ * Computes the inverse of the diagonal in region format and with interface scaling
+ */
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void jacobiSetup(RCP<Teuchos::ParameterList> params,
+                 const int maxRegPerProc,
+                 const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp,
+                 const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats,
+                 const Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionInterfaceScaling) {
+#include "Xpetra_UseShortNames.hpp"
+  const Scalar SC_ZERO = Teuchos::ScalarTraits<Scalar>::zero();
+  const Scalar SC_ONE = Teuchos::ScalarTraits<Scalar>::one();
+
+  Array<RCP<Vector> > regRes(maxRegPerProc);
+  createRegionalVector(regRes, maxRegPerProc, revisedRowMapPerGrp);
+
+  // extract diagonal from region matrices, recover true diagonal values, invert diagonal
+  Teuchos::Array<RCP<Vector> > diag(maxRegPerProc);
+  for (int j = 0; j < maxRegPerProc; j++) {
+    // extract inverse of diagonal from matrix
+    diag[j] = VectorFactory::Build(regionGrpMats[j]->getRowMap(), true);
+    regionGrpMats[j]->getLocalDiagCopy(*diag[j]);
+    diag[j]->elementWiseMultiply(SC_ONE, *regionInterfaceScaling[j], *diag[j], SC_ZERO); // ToDo Does it work to pass in diag[j], but also return into the same variable?
```

Comment:
Regarding the question in your comment: It works just fine :-)  In general, if there are no cross-loop data dependencies in the obvious `#pragma omp parallel for` implementation, you're fine.